### PR TITLE
Extract JVM Toolchain plugins to new JVM Platform project

### PR DESCRIPTION
**Pull request by tresat**
This PR extracts the `JvmToolchainsPlugin` and `JvmToolchainManagementPlugin` into a new `toolchains-jvm` project that lives under the JVM platform.

As a part of this, there are many resources that were in the `plugins` project that have been relocated to test fixtures the `language-groovy` project in order to be accessible to tests in both `plugins` and `toolchains-jvm`.  As these fixtures use resource files, the `Resources` class has been improved to be able to extract resources from within jars, so that it can extract them from the `test-fixtures` jars that will now be present on the test runtime classpath.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/internal-testing/src/main/groovy/org/gradle/util/internal/Resources.java [hunk 1]
```diff
@@ -45,22 +61,89 @@ public TestFile getResource(String name) {
...
+
+        try (JarFile sourceJar = new JarFile(sourceJarPath)) {
+            for (JarEntry sourceJarEntry : Collections.list(sourceJar.entries())) {
+                File currFile = new File(destDir, sourceJarEntry.getName());
+                if (sourceJarEntry.isDirectory()) {
+                    if (!currFile.mkdir()) {
+                        throw new IOException("Could not create directory " + currFile);
+                    }
+                } else {
+                    try (InputStream inputStream = sourceJar.getInputStream(sourceJarEntry);
```
Comment (big-guy) [inline subprojects/internal-testing/src/main/groovy/org/gradle/util/internal/Resources.java:138]: IIUC, `Files.copy(...)` might work here

File: subprojects/internal-testing/src/main/groovy/org/gradle/util/internal/Resources.java [hunk 2]
```diff
@@ -45,22 +61,89 @@ public TestFile getResource(String name) {
...
+            File outputDir;
+
+            synchronized (extractedJars) {
+                if (!extractedJars.containsKey(testClass)) {
+                    extractedJars.put(testClass, new HashMap<>());
+                }
+
+                Map<String, File> extractedJarsForTestClass = extractedJars.get(testClass);
+                if (!extractedJarsForTestClass.containsKey(jarFilePath)) {
+                    outputDir = Files.createTempDir();
```
Comment (big-guy) [inline subprojects/internal-testing/src/main/groovy/org/gradle/util/internal/Resources.java:109]: I think we should try to find a way to extract this into the final destination vs copying it to the system level tmp directory first.

**Review by big-guy — CHANGES_REQUESTED** [review]

File: platforms/jvm/toolchains-jvm/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    integTestDistributionRuntimeOnly(project(":distributions-jvm"))
+}
+
+configurations {
+    integTestImplementation {
+        attributes {
+            attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, objects.named(LibraryElements::class.java, LibraryElements.CLASSES))
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (tresat) [reply platforms/jvm/toolchains-jvm/build.gradle.kts:72]: No, will remove.

File: subprojects/internal-integ-testing/build.gradle.kts [hunk 4]
```diff
@@ -47,6 +47,7 @@ dependencies {
     implementation(project(":launcher"))
     implementation(project(":build-events"))
     implementation(project(":build-option"))
+    implementation(project(":toolchains-jvm"))
```
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/internal-integ-testing/build.gradle.kts:50]: The `JavaInstallationRegistry` class was previously in the `jvm-services` subproject.  It is only currently used in production code by `ToolchainsJVMServices`, `JavaToolchainQueryService` and `ShowToolchainsTask` all moved to the new `toolchains-jvm` project.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: platforms/jvm/toolchains-jvm/src/main/java/org/gradle/internal/jvm/inspection/JavaInstallationRegistry.java
Comment (jvandort) [inline platforms/jvm/toolchains-jvm/src/main/java/org/gradle/internal/jvm/inspection/JavaInstallationRegistry.java:1]: Can we avoid moving this class out of `jvm-services`? It seems to be the only one which came from `jvm-services`, though this class cannot depend on `:dependency-management` and such if it is to be used in the launcher for `Toolchains for the Gradle daemon`

**Review by jvandort — CHANGES_REQUESTED** [review]

File: platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJVMServices.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.internal.services;
+
+import org.gradle.internal.service.ServiceRegistration;
+import org.gradle.internal.service.scopes.AbstractPluginServiceRegistry;
+import org.gradle.jvm.toolchain.internal.DefaultJavaToolchainService;
+
+public class ToolchainsJVMServices extends AbstractPluginServiceRegistry {
```
[the remark this answers is not in the document]
Comment (tresat) [reply platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJVMServices.java:48]: Sure, will rename.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test QFL

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test QFL

**Comment by tresat** [comment]
@bot-gradle test QFL

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJvmServices.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            return objectFactory.newInstance(DefaultJvmToolchainManagement.class, registry);
+        }
+
+        protected JdkCacheDirectory createJdkCacheDirectory(ObjectFactory objectFactory, GradleUserHomeDirProvider homeDirProvider, FileOperations operations, FileLockManager lockManager, JvmMetadataDetector detector) {
+            return objectFactory.newInstance(JdkCacheDirectory.class, homeDirProvider, operations, lockManager, detector);
+        }
+
+        protected JavaInstallationRegistry createJavaInstallationRegistry(ObjectFactory objectFactory, List<InstallationSupplier> suppliers, BuildOperationExecutor executor, OperatingSystem os) {
+            return objectFactory.newInstance(JavaInstallationRegistry.class, suppliers, executor, os);
+        }
```
Comment (big-guy) [inline platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJvmServices.java:74]: 🤔 I think all of these could be replaced by just adding them to the `registerBuildServices` below with `registration.add(...)`

**Review by big-guy — APPROVED** [review]

File: platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJvmServices.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    }
+
+    private void registerJavaInstallationSuppliers(ServiceRegistration registration) {
+        registration.add(AutoInstalledInstallationSupplier.class);
```
[the remark this answers is not in the document]
Comment (tresat) [reply platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJvmServices.java:79]: This is a mistake I think.  I'll restore this registration.

File: platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJvmServices.java (hunk 6 above)
Comment (tresat) [reply platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJvmServices.java:74]: This works for everything except for`DefaultJavaToolchainResolverRegistry`, which throws `InstantiationException` and can't be created.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/tooling-api-builders/build.gradle.kts [hunk 8]
```diff
@@ -25,6 +25,7 @@ dependencies {
     implementation(project(":testing-base"))
     implementation(project(":testing-jvm"))
     implementation(project(":tooling-api"))
+    implementation(project(":toolchains-jvm"))
```
Comment (jvandort) [inline subprojects/tooling-api-builders/build.gradle.kts:28]: What code here requires classes from `toolchains-jvm`? I can't find any files in this project which reference any of the toolchains classes

File: subprojects/internal-integ-testing/build.gradle.kts (hunk 4 above)
Comment (jvandort) [inline subprojects/internal-integ-testing/build.gradle.kts:50]: Is this dependency still required now that `JavaInstallationRegistry` is not being moved?

File: subprojects/docs/build.gradle [hunk 9]
```diff
@@ -620,13 +620,18 @@ tasks.named("docsTest") { task ->
...
 
-        // Do not execute Java Module tests on JVM < 9
-        if (!javaVersion.java9Compatible) {
-            excludeTestsMatching "org.gradle.docs.samples.*.java-modules-*.sample"
-            excludeTestsMatching "org.gradle.docs.samples.*.incubating-java-modules-*.sample"
-            excludeTestsMatching 'org.gradle.docs.samples.*.snippet-java-application*.sample'
-            excludeTestsMatching 'org.gradle.docs.samples.*.snippet-java-library-module*.sample'
-            excludeTestsMatching 'org.gradle.docs.samples.*.snippet-testing-patch-module*.sample'
+        /*
+         * Do not execute tests which include additional compiler warnings that cause output comparisons to fail on JDK 20:
```
Comment (jvandort) [inline subprojects/docs/build.gradle:624]: I think we should resolve this in a separate PR since it is not related to this change. 

Can we split this off into a separate PR in order to ensure we resolve this more quickly?

**Review by jvandort — CHANGES_REQUESTED** [review]

File: subprojects/docs/build.gradle (hunk 9 above)
Comment (tresat) [reply subprojects/docs/build.gradle:624]: This is just a workaround.  Bo said this was fixed now.  I'll try removing it after rebasing.

File: subprojects/plugins/src/testFixtures/groovy/org/gradle/groovy/compile/ApiGroovyCompilerIntegrationSpec.groovy
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/plugins/src/testFixtures/groovy/org/gradle/groovy/compile/ApiGroovyCompilerIntegrationSpec.groovy:1]: I could understand why having abstract test classes in test fixtures makes sense. That was not apparent by the name of the class. 

What do you think about prefixing this class with `Abstract` to make this more apparent?

File: subprojects/tooling-api-builders/build.gradle.kts (hunk 8 above)
Comment (jvandort) [reply subprojects/tooling-api-builders/build.gradle.kts:28]: Can we change this to a `runtimeOnly` dependency and add a `because`?

Comment (jvandort) [reply subprojects/tooling-api-builders/build.gradle.kts:28]: Or, perhaps there is a missing transitive dependency somewhere

File: subprojects/internal-integ-testing/build.gradle.kts (hunk 4 above)
Comment (tresat) [reply subprojects/internal-integ-testing/build.gradle.kts:50]: Looks like no.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/docs/build.gradle (hunk 9 above)
Comment (tresat) [reply subprojects/docs/build.gradle:624]: Not fixed in the latest master.

File: subprojects/tooling-api-builders/build.gradle.kts (hunk 8 above)
Comment (tresat) [reply subprojects/tooling-api-builders/build.gradle.kts:28]: So we can make this `testCompileOnly`.  I'll switch it to that.

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/plugins/src/testFixtures/groovy/org/gradle/groovy/compile/ApiGroovyCompilerIntegrationSpec.groovy
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/plugins/src/testFixtures/groovy/org/gradle/groovy/compile/ApiGroovyCompilerIntegrationSpec.groovy:1]: Seems like a good improvement to me.  Look at the hierarchy of that type though.  There a whole chain of abstract classes:
`AbstractToolchainGroovyCompileIntegrationTest` -> `ApiGroovyCompilerIntegrationSpec` -> `GroovyCompilerIntegrationSpec` -> `BasicGroovyCompilerIntegrationSpec` -> `MultiVersionIntegrationSpec` -> `AbstractIntegrationSpec`.  

If we rename one to add the prefix, we should rename them all.  I think it's worth doing, but I'll do this as a followup PR once this is merged.

File: subprojects/tooling-api-builders/build.gradle.kts (hunk 8 above)
Comment (jvandort) [reply subprojects/tooling-api-builders/build.gradle.kts]: This still seems a little fishy. I looked into it a bit more and it seems like this is required since `TestExecutionBuildTaskSchedulerTest` is `Mock`ing `Test`, which for some reason requires `JavaLauncher` on the classpath. Really, we should be defining `toolchains-jvm` as an `api` dependency for the project where `Test` lives. However, since we do not use `api` dependencies really, I don't think we should start here. 

I think what we have here is mostly fine, though I think its important to explain why we have dependencies declared when they're non-`implementation` dependencies. Can we either:
1. Change this back to `testImplementation` 
2. Add a `because`

Also, we should move this declaration down to other `test*` dependency declarations. We should not be co-locating `implementation` and `testImplementation` dependencies next to each other in the buildscript, for consistency.

**Review by jvandort — COMMENTED** [review]
Other than the nitpick comments here, I really think we should solve the circular `language-groovy` dependency cycle. Otherwise, this PR LGTM

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/docs/build.gradle (hunk 9 above)
Comment (tresat) [reply subprojects/docs/build.gradle]: This is removed now, as the latest in master has fixed this problem.

File: platforms/jvm/toolchains-jvm/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    testImplementation(testFixtures(project(":core")))
+    testImplementation(testFixtures(project(":language-groovy")))
+    testImplementation(testFixtures(project(":logging")))
+
+    testRuntimeOnly(project(":distributions-core")) {
+        because("Tests instantiate DefaultClassLoaderRegistry which requires a 'gradle-plugins.properties' through DefaultPluginModuleRegistry")
+    }
+
+    integTestImplementation(testFixtures(project(":model-core")))
+    integTestImplementation(testFixtures(project(":language-groovy")))
```
[the remark this answers is not in the document]
Comment (tresat) [reply platforms/jvm/toolchains-jvm/build.gradle.kts:59]: After some experimentation, here's what I think.

These fixtures should properly live in the `language-groovy` project, as that's where the `GroovyCompiler` types live.

`language-groovy` properly deps `toolchains-jvm` as an implementation dependency.

`plugins` deps the test fixtures of `language-groovy` so that integration tests can extend the base types contained in these fixtures.

Currently, `toolchains-jvm` also deps the test fixtures of `language-groovy`.  This isn't a problem for the build, but is kind of odd to see such a circular chain of dependencies.  This is due to the integration tests in the `toolchains-jvm` project testing groovy compilation scenarios and needing to extend the base types contained in these fixtures.

We could:

1. Move the extending integration tests to a new or existing project that deps both `toolchains-jvm` and `language-groovy`.  This would allow removing the `toolchains-jvm` dep on the test fixtures.  `integ-test` would work, but that's really just postponing dealing with the problem, since we want to divide that project into platforms and remove it.
2. Move the fixtures into a new or existing project that is depped by both `toolchains-jvm` and `language-groovy`.  Perhaps `platform-jvm`?  I don't like that this moves the fixtures away from the Groovy compiler types they are meant to test.
3. Move the tests out of `plugins` into `toolchains-jvm`.  Move the test fixtures from `language-groovy` into `toolchains-jvm`.  Now there is no need for anything to dep the `language-groovy` test fixtures, since they won't exist - everything will be located inside the `toolchains-jvm` project.  I think this is less correct, and bloats that project.
4. Do nothing
5. Do something else

Unless you have a good option 5, I favor option 4 as it seems to leave the classes where you would expect to find them.  I can add a `because` in the `toolchains-jvm` project's dep on the `language-groovy` test fixtures to lampshade the unusual dep cycle, but that cycle isn't hurting anything ATM.  Sterling and I did discuss test fixture cycles briefly and he was okay with it and approved this PR.  All else equal it would be better not to have it, but I think co-locating the tests and fixtures with the correct prod code is more important.  If this becomes a problem later it can be addressed then.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle cancel

**Comment by tresat** [comment]
@bot-gradle test and merge
